src/Ng/ElfieFromCatalogCollector.cs

```diff
@@ -339,35 +358,463 @@ string CreateIdxFile(string packageId, string packageVersion, string tempDirecto
+            //   3. Exclude any packages which do not have a latest stable version.
+            //   4. Sort the packages based on download count.
+            //   5. Include the most popular packages until the target download count is reached.
+            //   6. Sort the included packages, first by log2(downloadcount), then by package name.
+            //      This sort order ensures the most popular packages are recommended first, while
+            //      minimizing the day-to-day differences.
+
+
+            // The runninng count of the downloads included in the ardb.
+            long includedDownloadCount = 0;
+
+            // A flag to signal that the download threshold was reached.
+            bool thresholdReached = false;
+            List<Tuple<RegistrationIndexPackage, int>> includedPackagesWithCounts = new List<Tuple<RegistrationIndexPackage, int>>();
+
+            // We need to determine the latest stable version for each package. But this is a fairly expensive operation since
+            // it makes a network call. We'll process the packages in chuncks so we can get the latest stable version of
+            // the packages using multiple threads.
+            int batchSize = 500;
+            int currentPosition = 0;
+
+            // We need to process the packages in order from most popular to least popular. (item.Value is the download count for the package)
+            var orderedDownloadCounts = packageDownloadCounts.OrderByDescending(item => item.Value);
+            IEnumerable<string> batch = null;
+
+            do
+            {
+                // Get the next chunk of packages to process.
+                batch = orderedDownloadCounts.Skip(currentPosition).Take(batchSize).Select(item => item.Key);
+
+                // Get the latest stable versions of the packages.
+                Dictionary<string, RegistrationIndexPackage> latestStableVersions = GetLatestStableVersion(batch);
+
+                foreach (string packageId in batch)
+                {
+                    long downloadCount = packageDownloadCounts[packageId];
+                    RegistrationIndexPackage latestStableVersion;
+
+                    // If there's a latest stable version for the package, we want to include it.
+                    if (latestStableVersions.TryGetValue(packageId, out latestStableVersion))
+                    {
+                        Trace.TraceInformation($"Included package: {packageId} - {downloadCount.ToString("#,####")}");
+                        includedDownloadCount += downloadCount;
+
+                        // Calculate the log base 2 version of the download count. We'll use this value to
+                        // sort the packages in the ardb file.
+                        int base2Count = 0;
+                        if (downloadCount != 0)
+                        {
+                            base2Count = (int)Math.Log(downloadCount, 2);
```

Comment:
I'll add a comment for this in the code, but here's the explanation for the base2 download count.

Elfie is supposed to return suggestions in descending popularity order. This is because we believe user's are more likely to install the popular packages. However, if we order just by the download count, then the day-to-day deltas become large because packages will leapfrog each other (i.e. there are many packages that will move up or down one place in the sort order.) To alleviate this, we first group the packages by their log base 2 download count, and then alphabetically within each group. Now the packages can move within their log base 2 group and there's no day-to-day delta. We only see a delta when packages shift between log base 2 groups.